jhu-idc/idc-isle-dc
# Add S3 to IDC stack

Note:  This incorporates Drupal and dependency updates in #117 , so that PR should be merged first

## Overview

* Adds [s3fs](https://www.drupal.org/project/s3fs) module to Drupal
* Configures s3fs to provide implementation of `private://` filesystem in S3 (i.e. everything in the private filesystem is stored in S3 by `s3fs`)
* Configures migrations, media storage definitions, and derivatives to use private filesystem for all media.
* Adds a test to verify that media and derivatives truly are stored in S3.
* Adds `minio` service for local development stack as an S3 implementation.  This service is unnecessary in the cloud, as we'll just point to Amazon's S3 for S3 storage.

### Notable consequences

Because all binaries are in the `private://` filesystem, all access to them goes through Drupal.  In other words, with `public://`, access to files/binaries was provided by bare `nginx` serving the contents of the `sites/default/files` directory.  Now all hits to files (and thumbnails) will go through Drupal, which will dutifully stream content from S3.  
* There will be a negative performance impact, but it is unclear if this is significant or not.
* There may be implications on resource usage with larger file sizes, but it is unclear what that impact might be, and where (ram?  temporary space?  time?)
* With content now served by the private filesystem, there is an opportunity to control access based on roles.  By default, all content is open and can be accessed anonymously.   [certain modules](https://www.drupal.org/project/private_files_download_permission) can be used to restrict access, but are outside of the scope of this PR.

### Known issues

Currently, all content served from the private filesystem has `cache-control: private` header set, and an expiration date of 1978.  This means that no files or thumbnails are cacheable by browsers or proxies.    I tried using the [http_response_headers](https://www.drupal.org/project/http_response_headers) module to compensate, which `sucked`.  It is undocumented, and doesn't have half the features they claim in the README.  So if this PR is accepted, we'd need to file an issue for how to re-enable caching via http headers.  [this file](https://github.com/chapter-three/http_response_headers/blob/master/src/EventSubscriber/AddHTTPHeaders.php) gives a clue that this can be easily done in a module.

## To Test

Try out as many of the following challenges as you dare:

* Try uploading files, verify that nothing seems broken
* Try stopping minio (e.g. ` docker stop idc_minio_1`) and uploading an image, or performing a migration  Does it degrade "gracefully" enough?)
* I've manually pushed an s3-enabled `drupal-static` image:  `ghcr.io/jhu-sheridan-libraries/idc-isle-dc/drupal-static:upstream-20201007-739693ae-253-g4ee3114`.  Try running it on a cloud instance, [setting the proper environment variables](https://github.com/jhu-idc/idc-isle-dc/tree/issue-idc-general-26#configuration), and pointing it to a _bona fide_ bucket in AWS.  


Resolves #26
